render/fx_renderer/fx_pass.c

```diff
@@ -1015,6 +1026,219 @@ bool fx_render_pass_add_optimized_blur(struct fx_gles_render_pass *pass,
 	return !failed;
 }
 
+static void render_drop_shadow_blur_direction(struct fx_gles_render_pass *pass,
+		const struct fx_render_drop_shadow_options *fx_options,
+		bool is_horizontal) {
+	const struct wlr_render_texture_options *options = &fx_options->tex_options.base;
+	struct fx_renderer *renderer = pass->buffer->renderer;
+	struct fx_texture *texture = fx_get_texture(options->texture);
+	assert(texture->has_alpha);
+
+	struct drop_shadow_shader *shader = &renderer->shaders.drop_shadow;
+
+	struct wlr_box dst_box;
+	struct wlr_fbox src_fbox;
+	wlr_render_texture_options_get_src_box(options, &src_fbox);
+	wlr_render_texture_options_get_dst_box(options, &dst_box);
+
+	src_fbox.x /= options->texture->width;
+	src_fbox.y /= options->texture->height;
+	src_fbox.width /= options->texture->width;
+	src_fbox.height /= options->texture->height;
+
+	push_fx_debug(renderer);
+
+	// setup_blending(options->blend_mode);
+	setup_blending(WLR_RENDER_BLEND_MODE_NONE);
+	glBlendFunc(GL_ONE, GL_ZERO);
+
+	glUseProgram(shader->program);
+
+	glActiveTexture(GL_TEXTURE0);
+	glBindTexture(texture->target, texture->tex);
+
+	switch (options->filter_mode) {
+	case WLR_SCALE_FILTER_BILINEAR:
+		glTexParameteri(texture->target, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+		glTexParameteri(texture->target, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+		break;
+	case WLR_SCALE_FILTER_NEAREST:
+		abort();
+	}
+
+	glUniform1i(shader->tex, 0);
+	glUniform4f(shader->color, fx_options->color.r, fx_options->color.g, fx_options->color.b, fx_options->color.a);
+	glUniform1f(shader->blur_sigma, fx_options->blur_sigma);
+	glUniform2f(shader->direction, 1 - !is_horizontal, 1 - is_horizontal);
+	glUniform2f(shader->size, dst_box.width, dst_box.height);
+
+	set_proj_matrix(shader->proj, pass->projection_matrix, &dst_box);
+	set_tex_matrix(shader->tex_proj, options->transform, &src_fbox);
+
+	render(&dst_box, options->clip, shader->pos_attrib);
+
+	glBindTexture(texture->target, 0);
+
+	glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+
+	pop_fx_debug(renderer);
+}
+
+static void fx_render_pass_add_drop_shadow_final(struct fx_gles_render_pass *pass,
+		const struct fx_render_drop_shadow_options *fx_options) {
```

Comment:
I'm wondering if instead we just pass the color and base options, since that's all we use.